test/e2e-encryption/encryption_test.go

```diff
@@ -1,23 +1,211 @@
+
+	apiServer, err := apiServerClient.Get("cluster", metav1.GetOptions{})
+	require.NoError(t, err)
+	apiServer.Spec.Encryption.Type = encryptionType
+	_, err = apiServerClient.Update(apiServer)
+	require.NoError(t, err)
+
+	test.WaitForKubeAPIServerClusterOperatorAvailableNotProgressingNotDegraded(t, configClient)
+
+	return kv, done
+}
+
+func getEncryptionClients(t *testing.T) (test.EtcdGetter, func(), configv1client.ConfigV1Interface, configv1client.APIServerInterface, kubernetes.Interface, v1helpers.StaticPodOperatorClient) {
+	t.Helper()
+
 	kubeConfig, err := test.NewClientConfigForTest()
 	require.NoError(t, err)
-	kubeClient, err := kubernetes.NewForConfig(kubeConfig)
+
+	configClient := configv1client.NewForConfigOrDie(kubeConfig)
+	apiServerClient := configClient.APIServers()
+
+	kubeClient := kubernetes.NewForConfigOrDie(kubeConfig)
+
+	kv := test.NewEtcdGetter(kubeClient)
+
+	gvr := operatorv1.GroupVersion.WithResource("kubeapiservers")
+	operatorClient, dynamicInformers, err := genericoperatorclient.NewStaticPodOperatorClient(kubeConfig, gvr)
 	require.NoError(t, err)
-	kv, done := test.NewEtcdKVMust(t, kubeClient)
+	stopCh := make(chan struct{})
+	dynamicInformers.Start(stopCh)
+
+	timeout, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+	require.Truef(t, dynamicInformers.WaitForCacheSync(timeout.Done())[gvr], "failed to sync cache for %s", gvr)
+
+	done := func() {
+		close(stopCh)
+	}
+
+	return kv, done, configClient, apiServerClient, kubeClient, operatorClient
+}
+
+func TestEncryptionRotation(t *testing.T) {
+	kv, done, configClient, apiServerClient, kubeClient, operatorClient := getEncryptionClients(t)
 	defer done()
+	secretsClient := kubeClient.CoreV1().Secrets(operatorclient.GlobalMachineSpecifiedConfigNamespace)
+
+	apiServer, err := apiServerClient.Get("cluster", metav1.GetOptions{})
+	require.NoError(t, err)
+	apiServer.Spec.Encryption.Type = configv1.EncryptionTypeAESCBC
+	_, err = apiServerClient.Update(apiServer)
+	require.NoError(t, err)
+
+	secretsPrefixes := make([]string, 3)
+	cmPrefixes := make([]string, 3)
+
+	// run a few rotations and assert that migrations occur as expected and keyIDs increase
+	for i := range secretsPrefixes {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			secretsKeyPrefix, cmKeyPrefix := testRotation(t, operatorClient, secretsClient, configClient, kv)
+			secretsPrefixes[i] = secretsKeyPrefix
+			cmPrefixes[i] = cmKeyPrefix
+		})
+		if t.Failed() {
+			return
+		}
+	}
+
+	require.Truef(t, sort.IsSorted(sort.StringSlice(secretsPrefixes)), "secret key IDs not in ascending order: %v", secretsPrefixes)
+	require.Truef(t, sort.IsSorted(sort.StringSlice(cmPrefixes)), "config map key IDs not in ascending order: %v", cmPrefixes)
+}
+
+func testRotation(t *testing.T, operatorClient v1helpers.StaticPodOperatorClient, secretsClient corev1.SecretInterface, configClient configv1client.ConfigV1Interface, kv test.EtcdGetter) (string, string) {
```

Comment:
Is validating annotations and prefixes the most reliable way of checking if the content was encrypted?

Wouldn't it be better to validate the content? In this case, rotation must be reflected in the content change.

Reply:
I will explore this in a follow-up.